javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,284 @@
+  }
+
+  override def clear(): Unit =
+    removeRange(0, size)
+
+  def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+    checkIndexOnBounds(index)
+    for ((elem, i) <- c.iterator().zipWithIndex)
+      add(index + i, elem)
+    c.nonEmpty
+  }
+
+  def iterator(): Iterator[E] =
+    listIterator()
+
+  def listIterator(): ListIterator[E] =
+    listIterator(0)
+
+  def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new RandomAccessListIterator(index, 0, size)
+  }
+
+  def subList(fromIndex: Int, toIndex: Int): List[E] = {
+    if (fromIndex < 0 || toIndex > size)
+      throw new IndexOutOfBoundsException
+    else if (fromIndex > toIndex)
+      throw new IllegalArgumentException
+
+    self match {
+      case _: RandomAccess =>
+        new AbstractListView(fromIndex, toIndex) with RandomAccess {
+          override def listIterator(index: Int): ListIterator[E] = {
+            checkIndexOnBounds(index)
+            // Iterator that accesses the original list directly
+            new RandomAccessListIterator(fromIndex + index, fromIndex, toIndex)
+          }
+        }
+      case _ =>
+        new AbstractListView(fromIndex, toIndex) {
+          override def listIterator(index: Int): ListIterator[E] = {
+            checkIndexOnBounds(index)
+            // Iterator that accesses the original list using it's iterator
+            new BackedUpListIterator(index, fromIndex)
+          }
+        }
+    }
+  }
+
+  override def equals(o: Any): Boolean = {
+    if (o.asInstanceOf[AnyRef] eq this) {
+      true
+    } else {
+      o match {
+        case o: List[_] =>
+          val oIter = o.listIterator
+          this.forall(oIter.hasNext && _ === oIter.next) && !oIter.hasNext
+        case _ => false
+      }
+    }
+  }
+
+  override def hashCode(): Int = {
+    this.foldLeft(1) {
+      (prev, elem) => 31 * prev + (if (elem == null) 0 else elem.hashCode)
+    }
+  }
+
+  protected def removeRange(fromIndex: Int, toIndex: Int): Unit = {
+    @tailrec
+    def removeNext(index: Int, iter: ListIterator[E]): Unit = {
+      if (iter.hasNext && index < toIndex) {
+        println(iter.next())
+        iter.remove()
+        removeNext(index + 1, iter)
+      }
+    }
+    removeNext(fromIndex, listIterator(fromIndex))
+  }
+
+  protected[this] def checkIndexInBounds(index: Int): Unit = {
+    if (index < 0 || index >= size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  protected[this] def checkIndexOnBounds(index: Int): Unit = {
+    if (index < 0 || index > size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  private abstract class AbstractListView(fromIndex: Int,
+        private var toIndex: Int) extends AbstractList[E] {
+
+    override def add(e: E): Boolean = {
+      add(size, e)
+      toIndex += 1
+      true
+    }
+
+    override def add(index: Int, e: E): Unit = {
+      checkIndexOnBounds(index)
+      self.add(fromIndex + index, e)
```

Comment:
Shouldn't this increment `toIndex`?

Consequence: `add(e: E)` shouldn't do it itself, and hence need not even be overridden at all.